File: src/csharp/pulse/cdm/scalars/SEScalarVolume.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class VolumeUnit : Unit
  {
    private VolumeUnit(string v) { Value = v; }

    public static VolumeUnit L = new VolumeUnit("L");
    public static VolumeUnit mL = new VolumeUnit("mL");
    public static VolumeUnit uL = new VolumeUnit("uL");
    public static VolumeUnit dL = new VolumeUnit("dL");
    public static VolumeUnit m3 = new VolumeUnit("m^3");

    public static VolumeUnit FromString(string u)
    {
      if (u == VolumeUnit.L.ToString())
        return VolumeUnit.L;
      if (u == VolumeUnit.mL.ToString())
        return VolumeUnit.mL;
      if (u == VolumeUnit.uL.ToString())
        return VolumeUnit.uL;
      if (u == VolumeUnit.dL.ToString())
        return VolumeUnit.dL;
      if (u == VolumeUnit.m3.ToString())
        return VolumeUnit.m3;

      throw new System.ArgumentException(u + " is not a know volume unit");
    }
  }

  public class SEScalarVolume : SEScalar
  {
    protected VolumeUnit unit;

    public SEScalarVolume() : base()
    {

    }

    public SEScalarVolume(SEScalarVolume from)
    {
      Set(from);
    }

    public SEScalarVolume(double value, VolumeUnit unit)
    {
      SetValue(value, unit);
    }

    public void Set(SEScalarVolume from)
    {
      if (from == null)
      {
        Invalidate();
        return;
      }
      SetValue(from.value, from.unit);
    }

    public new void SetValue(double value)
    {
      throw new System.NotSupportedException("You must provide a unit");
    }

    public void SetValue(double value, VolumeUnit unit)
    {
      this.value = value;
      this.unit = unit;
    }

    public double GetValue(VolumeUnit unit)
    {
      return UnitConverter.Convert(this.value, this.unit, unit);
    }

    public VolumeUnit GetUnit()
    {
      return unit;
    }

    public new string ToString()
    {
      return this.value + "(" + this.unit.ToString() + ")";
    }
  }
}

File: src/cpp/cdm/properties/SEScalarVolume.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarVolume.h"

const VolumeUnit VolumeUnit::L("L");
const VolumeUnit VolumeUnit::dL("dL");
const VolumeUnit VolumeUnit::mL("mL");
const VolumeUnit VolumeUnit::uL("uL");
const VolumeUnit VolumeUnit::m3("m^3");

bool VolumeUnit::IsValidUnit(const std::string& unit)
{
  if (L.GetString().compare(unit) == 0)
    return true;
  if (dL.GetString().compare(unit) == 0)
    return true;
  if (mL.GetString().compare(unit) == 0)
    return true;
  if (uL.GetString().compare(unit) == 0)
    return true;
  if (m3.GetString().compare(unit) == 0)
    return true;
  return false;
}

const VolumeUnit& VolumeUnit::GetCompoundUnit(const std::string& unit)
{
  if(L.GetString().compare(unit)==0)
    return L;
  if(mL.GetString().compare(unit)==0)
    return mL;
  if(m3.GetString().compare(unit)==0)
    return m3;
  if (dL.GetString().compare(unit) == 0)
    return dL;
  if (uL.GetString().compare(unit) == 0)
    return uL;
  throw CommonDataModelException(unit + " is not a valid Volume unit");
}

template class SEScalarQuantity<VolumeUnit>;
